### PR TITLE
[ggj][codegen](1/2)feat: Add equals() method in ResourceName

### PR DESCRIPTION
**Pull request by summer-ji-eng**
sample code:
[java-logging](https://github.com/googleapis/java-logging/blob/master/proto-google-cloud-logging-v2/src/main/java/com/google/logging/v2/FolderName.java#L143-L153
)
[java-pubsub](https://github.com/googleapis/java-pubsub/blob/master/proto-google-cloud-pubsub-v1/src/main/java/com/google/pubsub/v1/TopicName.java#L193-L204)

### CODE REVIEW

File: src/main/java/com/google/api/generator/gapic/composer/ResourceNameHelperClassComposer.java [hunk 1]
```diff
@@ -1140,6 +1145,113 @@ private static MethodDefinition createToStringMethod(
...
+    TypeNode thisClassType = types.get(getThisClassName(resourceName));
+    ValueExpr thisValueExpr = ValueExpr.withValue(ThisObjectValue.withType(thisClassType));
+    ValueExpr trueValueExpr =
+        ValueExpr.withValue(
+            PrimitiveValue.builder().setType(TypeNode.BOOLEAN).setValue("true").build());
+    ValueExpr falseValueExpr =
+        ValueExpr.withValue(
+            PrimitiveValue.builder().setType(TypeNode.BOOLEAN).setValue("false").build());
+
+    ReturnExpr returnTrueExpr = ReturnExpr.withExpr(trueValueExpr);
```
Comment (miraleung) [inline src/main/java/com/google/api/generator/gapic/composer/ResourceNameHelperClassComposer.java:1164]: Add some comments per section?

File: src/main/java/com/google/api/generator/gapic/composer/ResourceNameHelperClassComposer.java [hunk 2]
```diff
@@ -1140,6 +1145,113 @@ private static MethodDefinition createToStringMethod(
...
+            getClassMethodInvocationExpr
+                .toBuilder()
+                .setExprReferenceExpr(argVarExprNotDecl)
+                .build());
+    LogicalOperationExpr orLogicalExpr =
+        LogicalOperationExpr.logicalOrWithExprs(oNotEqualsNullExpr, getClassEqualsExpr);
+
+    Variable thatVariable = Variable.builder().setName("that").setType(thisClassType).build();
+    VariableExpr thatVariableExpr =
+        VariableExpr.builder().setIsDecl(true).setVariable(thatVariable).build();
```
Comment (miraleung) [inline src/main/java/com/google/api/generator/gapic/composer/ResourceNameHelperClassComposer.java:1184]: Could we not set this property, and instead use `toBuilder().setIsDecl(true).toBuild()` where needed?

**Review by miraleung — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/api/generator/gapic/composer/ResourceNameHelperClassComposer.java [hunk 3]
```diff
@@ -1140,6 +1145,113 @@ private static MethodDefinition createToStringMethod(
         .build();
   }
 
+  private static MethodDefinition createEqualsMethod(
+      ResourceName resourceName, List<List<String>> tokenHierarchies, Map<String, TypeNode> types) {
+    Variable oVariable = Variable.builder().setType(TypeNode.OBJECT).setName("o").build();
+    VariableExpr argVarExprDecl =
```
[the remark this answers is not in the document]
Comment (summer-ji-eng) [reply src/main/java/com/google/api/generator/gapic/composer/ResourceNameHelperClassComposer.java:1151]: sg

File: src/main/java/com/google/api/generator/gapic/composer/ResourceNameHelperClassComposer.java (hunk 2 above)
Comment (summer-ji-eng) [reply src/main/java/com/google/api/generator/gapic/composer/ResourceNameHelperClassComposer.java:1184]: sg

File: src/main/java/com/google/api/generator/gapic/composer/ResourceNameHelperClassComposer.java (hunk 1 above)
Comment (summer-ji-eng) [reply src/main/java/com/google/api/generator/gapic/composer/ResourceNameHelperClassComposer.java:1161]: sure

**Review by miraleung — APPROVED** [review]

**Pull request merged**